llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Utils/Utils.cpp

```diff
@@ -92,6 +92,46 @@ SmallVector<int64_t> asShapeWithAnyValueAsDynamic(ArrayRef<OpFoldResult> ofrs) {
   return result;
 }
 
+FailureOr<SmallVector<OpFoldResult>>
+getInnerTileSizesOfr(OpBuilder &rewriter, Location loc,
+                     RankedTensorType tensorType,
+                     const MaterializeEncodingInfo &materializeEncodingInfo,
+                     MaterializeEncodingValueFn materializeEncodingValueFn) {
+  ArrayRef<int64_t> staticTileSizes = materializeEncodingInfo.innerTileSizes;
+  if (llvm::all_of(staticTileSizes,
+                   [](int64_t i) { return !ShapedType::isDynamic(i); })) {
+    return SmallVector<OpFoldResult>(
+        llvm::map_range(staticTileSizes, [&](int64_t v) {
+          return rewriter.getI64IntegerAttr(v);
+        }));
+  }
+  assert(materializeEncodingValueFn &&
+         "When dynamic tile sizes are generated, a MaterializeEncodingValueFn "
+         "should be provided.");
+
+  FailureOr<MaterializeEncodingValueInfo> materializeEncodingValueInfo =
+      materializeEncodingValueFn(tensorType, rewriter, loc);
+  if (failed(materializeEncodingValueInfo)) {
+    return failure();
+  }
+  ArrayRef<Value> innerTileSizeValues =
+      materializeEncodingValueInfo->innerTileSizes;
+
+  SmallVector<OpFoldResult> result(staticTileSizes.size());
+  for (size_t i = 0; i < result.size(); ++i) {
+    if (staticTileSizes[i] == ShapedType::kDynamic) {
+      result[i] = innerTileSizeValues[i];
+    } else if (tensorType.isDynamicDim(i)) {
+      result[i] =
+          rewriter.create<arith::ConstantIndexOp>(loc, staticTileSizes[i])
+              .getResult();
+    } else {
+      result[i] = rewriter.getI64IntegerAttr(staticTileSizes[i]);
+    }
+  }
+  return result;
```

Comment:
Why do we pass `tensorType` to the function? Can't this just return `innerTileSizeValues`?

Reply:
as discussed - the function needs the type shape and encoding. The reason why it needs the shape is to perform the adjustment to narrow cases. Passing the tensorType is shorthand for that,  is consistent with the signature of the other callback (MaterializeEncodingFn), and gives us room to adjust what info exactly the function needs to use without changing interfaces.